src/utils.c

```diff
@@ -2499,149 +2499,142 @@ gchar *utils_get_os_info_string(void)
-			executable = utils_remove_ext_from_filename(doc->file_name);
-			replacement = g_path_get_basename(executable);
-			utils_string_replace_all(haystack, "%e", replacement);
-			g_free(replacement);
-			g_free(executable);
-		}
-
-		if (g_utf8_strchr(needles, -1, 'f') != NULL) {
-			/* replace %f with the filename (including extension) */
-			replacement = g_path_get_basename(doc->file_name);
-			utils_string_replace_all(haystack, "%f", replacement);
-			g_free(replacement);
-		}
-
-		if (g_utf8_strchr(needles, -1, 'l') != NULL) {
-			/* replace %l with the current 1-based line number */
-			line_num = sci_get_current_line(doc->editor->sci) + 1;
-			replacement = g_strdup_printf("%i", line_num);
-			utils_string_replace_all(haystack, "%l", replacement);
-			g_free(replacement);
+		p++;
+		if (! *p || *p == '%') /* %% or trailing % -> % */
+			g_string_append_c(res, '%');
+		else if (! insert_replacement(res, *p, data))
+		{
+			/* unknown placeholder, leave it literally */
+			g_string_append_c(res, '%');
+			g_string_append_c(res, *p);
 		}
 
-		/* if app->Project then                      */
-		/*   all %p placeholders are already removed */
-		/* else                                      */
-		/*   fall back to absolute path/name.ext     */
-		if (g_utf8_strchr(needles, -1, 'p') != NULL) {
-			/* replace %p with the absolute path/filename (including extension) */
-			utils_string_replace_all(haystack, "%p", doc->file_name);
-		}
+		str = p;
+		if (*str)
+			str++;
+		else
+			break;
 	}
+	g_string_append(res, str);
 
-	/* Show error message if placeholders should have been replaced but are not replaced */
-	errormsg = g_string_new("");
+	return g_string_free(res, FALSE);
+}
 
-	if (g_utf8_strchr(needles, -1, 'd') != NULL && strstr(haystack->str, "%d")) {
-		g_string_append(errormsg, " %%d");
-	}
 
-	if (g_utf8_strchr(needles, -1, 'e') != NULL && strstr(haystack->str, "%e")) {
-		g_string_append(errormsg, " %%e");
-	}
+static gboolean generate_document_replacements(GString *buffer,
+		const gchar placeholder, gpointer data)
+{
+	const GeanyDocument *doc = data;
+	gchar *r = NULL;
 
-	if (g_utf8_strchr(needles, -1, 'f') != NULL && strstr(haystack->str, "%f")) {
-		g_string_append(errormsg, " %%f");
+	/* %p is the project base path, if any */
+	if (placeholder == 'p' && app->project)
+		r = project_get_base_path();
+	else if (! doc || ! doc->file_name)
+	{
+		ui_set_statusbar(FALSE, _("failed to substitute %%%c: document has no path"), placeholder);
+		/* TODO: Shouldn't we rather return an empty string for known placeholders?
+		 *       That's not what the previous code did, though. */
 	}
-
-	if (g_utf8_strchr(needles, -1, 'l') != NULL && strstr(haystack->str, "%l")) {
-		g_string_append(errormsg, " %%l");
+	else
+	{
+		switch (placeholder)
+		{
+			case 'p': /* we only end up here if no project are open, so fallback on %d */
+				ui_set_statusbar(FALSE, _("failed to substitute %%p, no project active"));
```

Comment:
I wonder if the message could be misleading a bit because we *did* substitute the `%p` placeholder with `%d` and it is no error.

I think we could either rephrase the message to make it more informational (e.g. `no project active, substituted with %%d`) or just drop it completely.

Reply:
Good point, either sound good.

Reply:
I updated the status message not to suggest `%p` might be invalid or not handed